internal/checker/checker.go

```diff
@@ -16951,193 +16956,186 @@ func (b *KeyBuilder) WriteGenericTypeReferences(source *Type, target *Type, igno
+			b.writeNode(e.labeledDeclaration)
 		}
 	}
 	if readonly {
-		b.WriteByte('!')
+		b.writeByte('!')
 	}
-	return b.String()
+	return b.hash()
 }
 
-func getTypeAliasInstantiationKey(typeArguments []*Type, alias *TypeAlias) string {
+func getTypeAliasInstantiationKey(typeArguments []*Type, alias *TypeAlias) xxh3.Uint128 {
 	return getTypeInstantiationKey(typeArguments, alias, false)
 }
 
-func getTypeInstantiationKey(typeArguments []*Type, alias *TypeAlias, singleSignature bool) string {
-	var b KeyBuilder
-	b.WriteTypes(typeArguments)
-	b.WriteAlias(alias)
+func getTypeInstantiationKey(typeArguments []*Type, alias *TypeAlias, singleSignature bool) xxh3.Uint128 {
+	var b keyBuilder
+	b.writeTypes(typeArguments)
+	b.writeAlias(alias)
 	if singleSignature {
-		b.WriteByte('!')
+		b.writeByte('!')
 	}
-	return b.String()
+	return b.hash()
 }
 
-func getIndexedAccessKey(objectType *Type, indexType *Type, accessFlags AccessFlags, alias *TypeAlias) string {
-	var b KeyBuilder
-	b.WriteType(objectType)
-	b.WriteByte(',')
-	b.WriteType(indexType)
-	b.WriteByte(',')
-	b.WriteUint64(uint64(accessFlags))
-	b.WriteAlias(alias)
-	return b.String()
+func getIndexedAccessKey(objectType *Type, indexType *Type, accessFlags AccessFlags, alias *TypeAlias) xxh3.Uint128 {
+	var b keyBuilder
+	b.writeType(objectType)
+	b.writeByte(',')
+	b.writeType(indexType)
+	b.writeByte(',')
+	hashWrite32(&b.h, accessFlags)
+	b.writeAlias(alias)
+	return b.hash()
 }
 
-func getTemplateTypeKey(texts []string, types []*Type) string {
-	var b KeyBuilder
-	b.WriteTypes(types)
-	b.WriteByte('|')
+func getTemplateTypeKey(texts []string, types []*Type) xxh3.Uint128 {
+	var b keyBuilder
+	b.writeTypes(types)
+	b.writeByte('|')
 	for i, s := range texts {
 		if i != 0 {
-			b.WriteByte(',')
+			b.writeByte(',')
 		}
-		b.WriteInt(len(s))
+		b.writeInt(len(s))
 	}
-	b.WriteByte('|')
+	b.writeByte('|')
 	for _, s := range texts {
-		b.WriteString(s)
+		b.writeString(s)
 	}
-	return b.String()
+	return b.hash()
 }
 
-func getConditionalTypeKey(typeArguments []*Type, alias *TypeAlias, forConstraint bool) string {
-	var b KeyBuilder
-	b.WriteTypes(typeArguments)
-	b.WriteAlias(alias)
+func getConditionalTypeKey(typeArguments []*Type, alias *TypeAlias, forConstraint bool) xxh3.Uint128 {
+	var b keyBuilder
+	b.writeTypes(typeArguments)
+	b.writeAlias(alias)
 	if forConstraint {
-		b.WriteByte('!')
+		b.writeByte('!')
 	}
-	return b.String()
+	return b.hash()
 }
 
-func getRelationKey(source *Type, target *Type, intersectionState IntersectionState, isIdentity bool, ignoreConstraints bool) string {
+func getRelationKey(source *Type, target *Type, intersectionState IntersectionState, isIdentity bool, ignoreConstraints bool) (xxh3.Uint128, bool) {
 	if isIdentity && source.id > target.id {
 		source, target = target, source
 	}
-	var b KeyBuilder
+	var b keyBuilder
 	var constrained bool
 	if isTypeReferenceWithGenericArguments(source) && isTypeReferenceWithGenericArguments(target) {
```

Comment:
I think we need something to change here. The old code relied on the fact that `writeGenericTypeReferences` would write strings that never start with a character that could occur in a `writeType` string. That's no longer the case because `writeType` now writes four bytes that could be anything. I think we need to write an explicit distinguishing prefix byte to indicate which branch we took.

Reply:
Added a byte before each.